builtin/logical/pki/acme_challenges.go

```diff
@@ -0,0 +1,128 @@
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	// Size of the raw and url-encoded base64 without padding thumbprint.
+	KeyAuthorizationThumbprintSizeBytes   = 32
+	KeyAuthorizationThumbprintSizeEncoded = 43
+
+	// Size of the raw and url-encoded base64 without padding token.
+	KeyAuthorizationTokenSizeBytes   = 32
+	KeyAuthorizationTokenSizeEncoded = 43
+)
+
+// We take this and apply it to all challenge methods for consistency.
+// Additionally, as the client controls generation of the thumbprint,
+func ValidateKeyAuthorization(keyAuthz string, token string, thumbprint string) (bool, error) {
+	parts := strings.Split(keyAuthz, ".")
+	if len(parts) != 2 {
+		return false, fmt.Errorf("invalid authorization: got %v parts, expected 2", len(parts))
+	}
+
+	tokenPart := parts[0]
+	thumbprintPart := parts[1]
+
+	if token != tokenPart || thumbprint != thumbprintPart {
+		return false, fmt.Errorf("key authorization was invalid")
+	}
+
+	return true, nil
+}
+
+func ValidateHTTP01Challenge(domain string, token string, thumbprint string) (bool, error) {
+	path := "http://" + domain + "/.well-known/acme-challenge/" + token
+
+	transport := &http.Transport{
+		// Only a single request is sent to this server as we do not do any
+		// batching of validation attempts. There is no need to do an HTTP
+		// KeepAlive as a result.
+		DisableKeepAlives:   true,
+		MaxIdleConns:        1,
+		MaxIdleConnsPerHost: 1,
+		MaxConnsPerHost:     1,
+		IdleConnTimeout:     1 * time.Second,
+
+		// We'd rather timeout and re-attempt validation later than hang
+		// too many validators waiting for slow hosts.
+		DialContext: (&net.Dialer{
+			Timeout:   10 * time.Second,
+			KeepAlive: -1 * time.Second,
+		}).DialContext,
+		ResponseHeaderTimeout: 10 * time.Second,
+	}
+
+	maxRedirects := 10
+	urlLength := 1024
+
+	client := &http.Client{
+		Transport: transport,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			if len(via)+1 >= maxRedirects {
+				return fmt.Errorf("http-01: too many redirects: %v", len(via)+1)
+			}
+
+			reqUrlLen := len(req.URL.String())
+			if reqUrlLen > urlLength {
+				return fmt.Errorf("http-01: redirect url length too long: %v", reqUrlLen)
+			}
+
+			return nil
+		},
+	}
+
+	resp, err := client.Get(path)
+	if err != nil {
+		return false, fmt.Errorf("http-01: failed to fetch path %v: %w", path, err)
+	}
+
+	// We provision a buffer which allows for a variable size challenge, some
+	// whitespace, and a detection gap for too long of a message.
+	minExpected := KeyAuthorizationTokenSizeEncoded + 1 + KeyAuthorizationTokenSizeEncoded
+	maxExpected := 512
+	detectionBuffer := 2
+	offset := 0
+	rawBodyLen := maxExpected + detectionBuffer
+	rawBody := make([]byte, rawBodyLen)
+
+	defer resp.Body.Close()
+
+	// Attempt to read the body, but don't do so infinitely.
+	for i := 0; i < 10; i++ {
+		n, err := resp.Body.Read(rawBody[offset:rawBodyLen])
+		if n > 0 {
+			offset += n
+			rawBodyLen -= n
+
+			if offset > maxExpected {
```

Comment:
is necessary to error out if someone has ended their file with \r\n?  Like this is comparing to maxExpected, not rawBodyLen

Reply:
`maxExpected` is well above `minExpected` (43*2 +1 == 89 bytes), so any whitespace padding should already be accounted for. `maxExpected` is our threshhold to detect if we've gotten a completely bogus result we shouldn't even be _considering_ as a challenge type at all. 

The extra two bytes here (`rawBodyLen` vs `maxExpected`) is so that, if someone does send too long of a body (i.e., 513 or more bytes), we can detect that versus getting a 512-byte read and not knowing if there's more in the buffer (too many bytes) or not.

This is thus mostly defensive.

Reply:
Just curious why not use [io.LimitedReader](https://pkg.go.dev/io#LimitedReader) instead?